File: src/vn97/capability_activation.py
from __future__ import annotations

from dataclasses import dataclass
from enum import Enum
import hashlib
from pathlib import Path
import secrets
from typing import Protocol

from .capability_inventory import (
    MAX_HISTORY,
    MAX_STACK_DEPTH,
    CapabilityActivationError,
    CapabilityInventoryItem,
    CapabilityInventoryStore,
    InventoryCorruptionError,
    InventoryEvent,
    InventorySnapshot,
    _ID_RE,
    _REV_RE,
    _StoredActivation,
    _canon,
    _require_id,
    _require_sha,
    _validate_item,
    _validate_plan_identity,
)
from .capability_trust import (
    AdapterSpec,
    CapabilityTrustStore,
    CompatibilityPlan,
    CompatibilityProfile,
    SignatureVerifier,
    VerifiedCapability,
    plan_capability_compatibility,
    verify_staged_capability,
)


class ActivationConflictError(CapabilityActivationError):
    pass


class ActivationRecoveryRequired(CapabilityActivationError):
    pass


class VersionTransitionError(CapabilityActivationError):
    pass


class BackendTransactionState(str, Enum):
    PREPARED = "prepared"
    COMMITTED = "committed"
    ROLLED_BACK = "rolled_back"


@dataclass(frozen=True)
class BackendStatus:
    state: BackendTransactionState
    runtime_revision: str | None = None

    def __post_init__(self) -> None:
        if self.state is BackendTransactionState.COMMITTED:
            if self.runtime_revision is None or not _REV_RE.fullmatch(self.runtime_revision):
                raise ValueError("committed status requires valid runtime_revision")
        elif self.runtime_revision is not None:
            raise ValueError("non-committed status must not contain runtime_revision")


@dataclass(frozen=True)
class PreparedActivation:
    backend_id: str
    token: str
    artifact_sha256: str

    def __post_init__(self) -> None:
        from .capability_inventory import _SHA_RE, _TOKEN_RE

        if not _ID_RE.fullmatch(self.backend_id):
            raise ValueError("backend_id is invalid")
        if not _TOKEN_RE.fullmatch(self.token):
            raise ValueError("backend token is invalid")
        if not _SHA_RE.fullmatch(self.artifact_sha256):
            raise ValueError("artifact_sha256 is invalid")


class CapabilityActivationBackend(Protocol):
    @property
    def backend_id(self) -> str: ...

    def prepare(
        self,
        verified: VerifiedCapability,
        plan: CompatibilityPlan,
    ) -> PreparedActivation: ...

    def commit(self, token: str) -> str: ...

    def inspect(self, token: str) -> BackendStatus: ...

    def rollback(self, token: str) -> None: ...


def _plan_obj(plan: CompatibilityPlan) -> dict[str, object]:
    return {
        "capability_id": plan.capability_id,
        "capability_version": plan.capability_version,
        "disposition": plan.disposition.value,
        "package_sha256": plan.package_sha256,
        "profile_id": plan.profile_id,
        "profile_sha256": plan.profile_sha256,
        "publisher_key_id": plan.publisher_key_id,
        "runtime_api_version": plan.runtime_api_version,
        "sections": [
            {
                "adapter_id": section.adapter_id,
                "lossy": section.lossy,
                "role": section.role,
                "section_index": section.section_index,
                "source_format": section.source_format,
                "target_format": section.target_format,
            }
            for section in plan.sections
        ],
        "signature_sha256": plan.signature_sha256,
    }


def compatibility_plan_sha256(plan: CompatibilityPlan) -> str:
    return hashlib.sha256(_canon(_plan_obj(plan))).hexdigest()


def _pending_plan_identity(
    plan: CompatibilityPlan,
    verified: VerifiedCapability,
) -> dict[str, object]:
    source = verified.parsed.manifest.source
    return {
        "capability_id": plan.capability_id,
        "capability_version": plan.capability_version,
        "package_sha256": plan.package_sha256,
        "plan_sha256": compatibility_plan_sha256(plan),
        "profile_id": plan.profile_id,
        "profile_sha256": plan.profile_sha256,
        "publisher_key_id": plan.publisher_key_id,
        "runtime_api_version": plan.runtime_api_version,
        "signature_sha256": plan.signature_sha256,
        "source_license": source.license,
        "source_origin": source.origin,
        "source_sha256": source.source_sha256,
    }


def _activation_id(
    plan_identity: dict[str, object],
    backend_id: str,
    artifact_sha256: str,
    revision: str,
) -> str:
    return hashlib.sha256(
        _canon(
            {
                **plan_identity,
                "artifact_sha256": artifact_sha256,
                "backend_id": backend_id,
                "runtime_revision": revision,
            }
        )
    ).hexdigest()


class CapabilityActivationCoordinator:
    def __init__(self, store: CapabilityInventoryStore) -> None:
        self.store = store

    def inventory(self) -> InventorySnapshot:
        return self.store.load()

    def activate(
        self,
        verified: VerifiedCapability,
        signature_envelope: bytes,
        plan: CompatibilityPlan,
        profile: CompatibilityProfile,
        backend: CapabilityActivationBackend,
        *,
        stage_root: Path,
        trust_store: CapabilityTrustStore,
        verifier: SignatureVerifier,
        adapters: tuple[AdapterSpec, ...] = (),
        allow_lossy: bool = False,
        allow_same_version_replace: bool = False,
        allow_downgrade: bool = False,
    ) -> CapabilityInventoryItem:
        fresh = verify_staged_capability(
            verified.staged,
            signature_envelope,
            stage_root=stage_root,
            trust_store=trust_store,
            verifier=verifier,
        )
        if fresh != verified:
            raise ActivationConflictError("verified capability changed before activation")
        expected = plan_capability_compatibility(
            fresh,
            profile,
            adapters=adapters,
            allow_lossy=allow_lossy,
        )
        if plan != expected:
            raise ActivationConflictError("compatibility plan is stale or mismatched")
        if not _ID_RE.fullmatch(backend.backend_id):
            raise ValueError("backend_id is invalid")
        identity = _pending_plan_identity(plan, fresh)

        with self.store.locked() as root_fd:
            state = self.store._load(root_fd)
            if state.pending is not None:
                raise ActivationRecoveryRequired("inventory has unresolved transaction")
            stack = state.stacks.get(plan.capability_id, [])
            current = stack[-1].item if stack else None
            if current is not None:
                exact = (
                    current.package_sha256 == plan.package_sha256
                    and current.signature_sha256 == plan.signature_sha256
                    and current.profile_sha256 == plan.profile_sha256
                    and current.plan_sha256 == identity["plan_sha256"]
                    and current.backend_id == backend.backend_id
                )
                if exact:
                    return current
                if len(stack) >= MAX_STACK_DEPTH:
                    raise VersionTransitionError("activation stack depth limit reached")
                if plan.capability_version < current.capability_version and not allow_downgrade:
                    raise VersionTransitionError("capability downgrade is denied")
                if (
                    plan.capability_version == current.capability_version
                    and not allow_same_version_replace
                ):
                    raise VersionTransitionError("same-version replacement is denied")

            tx_id = hashlib.sha256(
                _canon(
                    {
                        "generation": state.generation,
                        "identity": identity,
                        "nonce": secrets.token_hex(16),
                        "operation": "activate",
                    }
                )
            ).hexdigest()
            state.pending = {
                "activation_id": None,
                "artifact_sha256": None,
                "backend_id": backend.backend_id,
                "backend_token": None,
                "capability_id": plan.capability_id,
                "generation_base": state.generation,
                "identity": identity,
                "operation": "activate",
                "phase": "reserved",
                "tx_id": tx_id,
            }
            self.store._write(root_fd, state)

        try:
            prepared = backend.prepare(fresh, plan)
        except Exception:
            self._clear_reserved(tx_id)
            raise
        if prepared.backend_id != backend.backend_id:
            self._clear_reserved(tx_id)
            raise ActivationConflictError("backend prepare returned mismatched backend_id")

        with self.store.locked() as root_fd:
            state = self.store._load(root_fd)
            pending = state.pending
            if not pending or pending["tx_id"] != tx_id or pending["phase"] != "reserved":
                try:
                    backend.rollback(prepared.token)
                finally:
                    raise ActivationRecoveryRequired(
                        "activation reservation changed before commit"
                    )
            pending = dict(pending)
            pending.update(
                {
                    "phase": "prepared",
                    "backend_token": prepared.token,
                    "artifact_sha256": prepared.artifact_sha256,
                }
            )
            state.pending = pending
            self.store._write(root_fd, state)

        try:
            revision = backend.commit(prepared.token)
            if not isinstance(revision, str) or not _REV_RE.fullmatch(revision):
                raise CapabilityActivationError("backend returned invalid runtime_revision")
        except Exception:
            try:
                backend.rollback(prepared.token)
            except Exception as rollback_exc:
                raise ActivationRecoveryRequired(
                    "activation needs recovery after commit failure"
                ) from rollback_exc
            self._clear_prepared(tx_id, prepared.token)
            raise
        return self._finalize_activation(tx_id, revision, "activate")

    def rollback(
        self,
        capability_id: str,
        backend: CapabilityActivationBackend,
    ) -> CapabilityInventoryItem | None:
        _require_id(capability_id, "capability_id")
        with self.store.locked() as root_fd:
            state = self.store._load(root_fd)
            if state.pending is not None:
                raise ActivationRecoveryRequired("inventory has unresolved transaction")
            stack = state.stacks.get(capability_id)
            if not stack:
                raise CapabilityActivationError("capability is not active")
            current = stack[-1]
            if current.item.backend_id != backend.backend_id:
                raise ActivationConflictError("rollback backend mismatch")
            tx_id = hashlib.sha256(
                _canon(
                    {
                        "activation_id": current.item.activation_id,
                        "generation": state.generation,
                        "nonce": secrets.token_hex(16),
                        "operation": "rollback",
                    }
                )
            ).hexdigest()
            state.pending = {
                "activation_id": current.item.activation_id,
                "artifact_sha256": current.item.artifact_sha256,
                "backend_id": backend.backend_id,
                "backend_token": current.backend_token,
                "capability_id": capability_id,
                "generation_base": state.generation,
                "identity": {},
                "operation": "rollback",
                "phase": "prepared",
                "tx_id": tx_id,
            }
            self.store._write(root_fd, state)
        try:
            backend.rollback(current.backend_token)
        except Exception as exc:
            raise ActivationRecoveryRequired("rollback requires recovery") from exc
        return self._finalize_rollback(tx_id, "rollback")

    def recover(
        self,
        backends: dict[str, CapabilityActivationBackend],
    ) -> InventorySnapshot:
        with self.store.locked() as root_fd:
            state = self.store._load(root_fd)
            pending = state.pending
            if pending is None:
                return state.snapshot()
            if pending["phase"] == "reserved":
                state.pending = None
                self.store._write(root_fd, state)
                return state.snapshot()
            backend = backends.get(pending["backend_id"])
            if backend is None or backend.backend_id != pending["backend_id"]:
                raise ActivationRecoveryRequired("required backend is unavailable")
            token = pending["backend_token"]
            tx_id = pending["tx_id"]
            operation = pending["operation"]

        status = backend.inspect(token)
        if operation == "activate":
            if status.state is BackendTransactionState.COMMITTED:
                self._finalize_activation(
                    tx_id,
                    status.runtime_revision,
                    "recover_activate",
                )
            elif status.state in {
                BackendTransactionState.PREPARED,
                BackendTransactionState.ROLLED_BACK,
            }:
                if status.state is BackendTransactionState.PREPARED:
                    try:
                        backend.rollback(token)
                    except Exception as exc:
                        raise ActivationRecoveryRequired(
                            "prepared activation rollback failed"
                        ) from exc
                self._clear_prepared(tx_id, token)
            else:
                raise ActivationRecoveryRequired("unsupported activation recovery state")
        else:
            if status.state is not BackendTransactionState.ROLLED_BACK:
                try:
                    backend.rollback(token)
                except Exception as exc:
                    raise ActivationRecoveryRequired("rollback recovery failed") from exc
            self._finalize_rollback(tx_id, "recover_rollback")
        return self.inventory()

    def _clear_reserved(self, tx_id: str) -> None:
        with self.store.locked() as root_fd:
            state = self.store._load(root_fd)
            if (
                state.pending
                and state.pending["tx_id"] == tx_id
                and state.pending["phase"] == "reserved"
            ):
                state.pending = None
                self.store._write(root_fd, state)

    def _clear_prepared(self, tx_id: str, token: str) -> None:
        with self.store.locked() as root_fd:
            state = self.store._load(root_fd)
            if (
                state.pending
                and state.pending["tx_id"] == tx_id
                and state.pending["backend_token"] == token
            ):
                state.pending = None
                self.store._write(root_fd, state)

    def _finalize_activation(
        self,
        tx_id: str,
        revision: str | None,
        action: str,
    ) -> CapabilityInventoryItem:
        if revision is None or not _REV_RE.fullmatch(revision):
            raise ActivationRecoveryRequired(
                "committed activation lacks valid runtime_revision"
            )
        with self.store.locked() as root_fd:
            state = self.store._load(root_fd)
            pending = state.pending
            if (
                not pending
                or pending["tx_id"] != tx_id
                or pending["operation"] != "activate"
                or pending["phase"] != "prepared"
            ):
                raise ActivationRecoveryRequired(
                    "activation transaction changed before finalization"
                )
            identity = _validate_plan_identity(
                pending["identity"],
                expected_capability_id=pending["capability_id"],
            )
            artifact = _require_sha(
                pending["artifact_sha256"],
                "pending artifact_sha256",
            )
            token = pending["backend_token"]
            activation_id = _activation_id(
                identity,
                pending["backend_id"],
                artifact,
                revision,
            )
            item = CapabilityInventoryItem(
                activation_id=activation_id,
                capability_id=identity["capability_id"],
                capability_version=identity["capability_version"],
                package_sha256=identity["package_sha256"],
                publisher_key_id=identity["publisher_key_id"],
                signature_sha256=identity["signature_sha256"],
                profile_id=identity["profile_id"],
                profile_sha256=identity["profile_sha256"],
                plan_sha256=identity["plan_sha256"],
                runtime_api_version=identity["runtime_api_version"],
                backend_id=pending["backend_id"],
                artifact_sha256=artifact,
                runtime_revision=revision,
                source_origin=identity["source_origin"],
                source_sha256=identity["source_sha256"],
                source_license=identity["source_license"],
            )
            _validate_item(item)
            stack = state.stacks.setdefault(item.capability_id, [])
            if len(stack) >= MAX_STACK_DEPTH:
                raise ActivationRecoveryRequired("activation stack became full")
            stack.append(_StoredActivation(item, token))
            state.generation += 1
            state.history.append(
                InventoryEvent(
                    generation=state.generation,
                    action=action,
                    capability_id=item.capability_id,
                    activation_id=item.activation_id,
                    package_sha256=item.package_sha256,
                )
            )
            state.history = state.history[-MAX_HISTORY:]
            state.pending = None
            self.store._write(root_fd, state)
            return item

    def _finalize_rollback(
        self,
        tx_id: str,
        action: str,
    ) -> CapabilityInventoryItem | None:
        with self.store.locked() as root_fd:
            state = self.store._load(root_fd)
            pending = state.pending
            if (
                not pending
                or pending["tx_id"] != tx_id
                or pending["operation"] != "rollback"
            ):
                raise ActivationRecoveryRequired(
                    "rollback transaction changed before finalization"
                )
            stack = state.stacks.get(pending["capability_id"])
            if not stack or stack[-1].item.activation_id != pending["activation_id"]:
                raise ActivationConflictError("active capability changed during rollback")
            removed = stack.pop()
            if not stack:
                state.stacks.pop(removed.item.capability_id, None)
            state.generation += 1
            state.history.append(
                InventoryEvent(
                    generation=state.generation,
                    action=action,
                    capability_id=removed.item.capability_id,
                    activation_id=removed.item.activation_id,
                    package_sha256=removed.item.package_sha256,
                )
            )
            state.history = state.history[-MAX_HISTORY:]
            state.pending = None
            self.store._write(root_fd, state)
            return stack[-1].item if stack else None

File: android/runtime/src/main/java/ai/vn97/runtime/CapabilityActivation.kt
package ai.vn97.runtime

import java.io.File
import java.nio.charset.StandardCharsets
import java.security.SecureRandom

enum class VN97BackendTransactionState {
    PREPARED,
    COMMITTED,
    ROLLED_BACK,
}

data class VN97BackendStatus(
    val state: VN97BackendTransactionState,
    val runtimeRevision: String? = null,
) {
    init {
        if (state == VN97BackendTransactionState.COMMITTED) {
            inventoryRequireRevision(
                runtimeRevision ?: throw IllegalArgumentException(
                    "committed status requires runtime_revision"
                ),
                "runtime_revision",
            )
        } else if (runtimeRevision != null) {
            throw IllegalArgumentException(
                "non-committed backend status must not contain runtime_revision"
            )
        }
    }
}

data class VN97PreparedActivation(
    val backendId: String,
    val token: String,
    val artifactSha256: String,
) {
    init {
        inventoryRequireId(backendId, "backend_id")
        inventoryRequireToken(token, "backend token")
        inventoryRequireSha(artifactSha256, "artifact_sha256")
    }
}

interface VN97CapabilityActivationBackend {
    val backendId: String

    fun prepare(
        verified: VN97VerifiedCapability,
        plan: VN97CompatibilityPlan,
    ): VN97PreparedActivation

    fun commit(token: String): String
    fun inspect(token: String): VN97BackendStatus
    fun rollback(token: String)
}

class VN97ActivationConflictException(message: String) :
    VN97CapabilityActivationException(message)

class VN97ActivationRecoveryRequired(message: String, cause: Throwable? = null) :
    VN97CapabilityActivationException(message, cause)

class VN97VersionTransitionException(message: String) :
    VN97CapabilityActivationException(message)

class VN97CapabilityActivationCoordinator(
    private val store: VN97CapabilityInventoryStore,
) {
    private val random = SecureRandom()

    fun inventory(): VN97InventorySnapshot = store.load()

    fun activate(
        verified: VN97VerifiedCapability,
        plan: VN97CompatibilityPlan,
        profile: VN97CompatibilityProfile,
        backend: VN97CapabilityActivationBackend,
        stageRoot: File,
        trustStore: VN97CapabilityTrustStore,
        adapters: List<VN97AdapterSpec> = emptyList(),
        allowLossy: Boolean = false,
        allowSameVersionReplace: Boolean = false,
        allowDowngrade: Boolean = false,
    ): VN97CapabilityInventoryItem {
        val fresh = VN97CapabilityTrustVerifier.verify(
            staged = verified.staged,
            stageRoot = stageRoot,
            trustStore = trustStore,
        )
        if (fresh != verified) {
            throw VN97ActivationConflictException(
                "verified capability changed before activation"
            )
        }
        val expected = VN97CapabilityCompatibility.plan(
            fresh,
            profile,
            adapters,
            allowLossy,
        )
        if (plan != expected) {
            throw VN97ActivationConflictException(
                "compatibility plan is stale or mismatched"
            )
        }
        inventoryRequireId(backend.backendId, "backend_id")
        val identity = planIdentity(plan, fresh)

        var existing: VN97CapabilityInventoryItem? = null
        val txId = store.locked { tx ->
            val state = tx.state
            if (state.pending != null) {
                throw VN97ActivationRecoveryRequired(
                    "inventory has unresolved transaction"
                )
            }
            val stack = state.stacks[plan.capabilityId]
            val current = stack?.lastOrNull()?.item
            if (current != null) {
                val exact =
                    current.packageSha256 == plan.packageSha256 &&
                        current.signatureSha256 == plan.signatureSha256 &&
                        current.profileSha256 == plan.profileSha256 &&
                        current.planSha256 == identity.planSha256 &&
                        current.backendId == backend.backendId
                if (exact) {
                    existing = current
                    return@locked ""
                }
                if (stack.size >= 64) {
                    throw VN97VersionTransitionException(
                        "activation stack depth limit reached"
                    )
                }
                if (plan.capabilityVersion < current.capabilityVersion &&
                    !allowDowngrade
                ) {
                    throw VN97VersionTransitionException(
                        "capability downgrade is denied"
                    )
                }
                if (plan.capabilityVersion == current.capabilityVersion &&
                    !allowSameVersionReplace
                ) {
                    throw VN97VersionTransitionException(
                        "same-version replacement is denied"
                    )
                }
            }

            val id = activationTransactionId(
                generation = state.generation,
                identity = identity,
            )
            state.pending = VN97PendingTransaction(
                txId = id,
                operation = "activate",
                phase = "reserved",
                capabilityId = plan.capabilityId,
                backendId = backend.backendId,
                generationBase = state.generation,
                identity = identity,
                backendToken = null,
                artifactSha256 = null,
                activationId = null,
            )
            tx.persist()
            id
        }
        existing?.let { return it }

        val prepared = try {
            backend.prepare(fresh, plan)
        } catch (exc: Exception) {
            clearReserved(txId)
            throw exc
        }
        if (prepared.backendId != backend.backendId) {
            clearReserved(txId)
            throw VN97ActivationConflictException(
                "backend prepare returned mismatched backend_id"
            )
        }

        store.locked { tx ->
            val pending = tx.state.pending
            if (pending == null ||
                pending.txId != txId ||
                pending.phase != "reserved"
            ) {
                try {
                    backend.rollback(prepared.token)
                } finally {
                    throw VN97ActivationRecoveryRequired(
                        "activation reservation changed before commit"
                    )
                }
            }
            val currentPending = checkNotNull(tx.state.pending)
            tx.state.pending = currentPending.copy(
                phase = "prepared",
                backendToken = prepared.token,
                artifactSha256 = prepared.artifactSha256,
            )
            tx.persist()
        }

        val revision = try {
            backend.commit(prepared.token).also {
                inventoryRequireRevision(it, "runtime_revision")
            }
        } catch (exc: Exception) {
            try {
                backend.rollback(prepared.token)
            } catch (rollbackExc: Exception) {
                throw VN97ActivationRecoveryRequired(
                    "activation needs recovery after commit failure",
                    rollbackExc,
                )
            }
            clearPrepared(txId, prepared.token)
            throw exc
        }

        return finalizeActivation(
            txId = txId,
            revision = revision,
            action = VN97InventoryAction.ACTIVATE,
        )
    }

    fun rollback(
        capabilityId: String,
        backend: VN97CapabilityActivationBackend,
    ): VN97CapabilityInventoryItem? {
        inventoryRequireId(capabilityId, "capability_id")
        inventoryRequireId(backend.backendId, "backend_id")
        lateinit var token: String
        val txId = store.locked { tx ->
            val state = tx.state
            if (state.pending != null) {
                throw VN97ActivationRecoveryRequired(
                    "inventory has unresolved transaction"
                )
            }
            val stack = state.stacks[capabilityId]
                ?: throw VN97CapabilityActivationException(
                    "capability is not active"
                )
            val current = stack.last()
            if (current.item.backendId != backend.backendId) {
                throw VN97ActivationConflictException(
                    "rollback backend mismatch"
                )
            }
            token = current.backendToken
            val id = rollbackTransactionId(
                generation = state.generation,
                activationId = current.item.activationId,
            )
            state.pending = VN97PendingTransaction(
                txId = id,
                operation = "rollback",
                phase = "prepared",
                capabilityId = capabilityId,
                backendId = backend.backendId,
                generationBase = state.generation,
                identity = null,
                backendToken = token,
                artifactSha256 = current.item.artifactSha256,
                activationId = current.item.activationId,
            )
            tx.persist()
            id
        }

        try {
            backend.rollback(token)
        } catch (exc: Exception) {
            throw VN97ActivationRecoveryRequired(
                "rollback requires recovery",
                exc,
            )
        }
        return finalizeRollback(txId, VN97InventoryAction.ROLLBACK)
    }

    fun recover(
        backends: Map<String, VN97CapabilityActivationBackend>,
    ): VN97InventorySnapshot {
        val pending = store.locked { tx ->
            val value = tx.state.pending ?: return@locked null
            if (value.phase == "reserved") {
                tx.state.pending = null
                tx.persist()
                return@locked null
            }
            value
        } ?: return inventory()

        val backend = backends[pending.backendId]
            ?: throw VN97ActivationRecoveryRequired(
                "required recovery backend is unavailable"
            )
        if (backend.backendId != pending.backendId) {
            throw VN97ActivationRecoveryRequired(
                "required recovery backend identity mismatch"
            )
        }
        val token = pending.backendToken
            ?: throw VN97ActivationRecoveryRequired(
                "prepared transaction lacks backend token"
            )
        val status = backend.inspect(token)

        if (pending.operation == "activate") {
            when (status.state) {
                VN97BackendTransactionState.COMMITTED -> finalizeActivation(
                    txId = pending.txId,
                    revision = checkNotNull(status.runtimeRevision),
                    action = VN97InventoryAction.RECOVER_ACTIVATE,
                )
                VN97BackendTransactionState.PREPARED -> {
                    try {
                        backend.rollback(token)
                    } catch (exc: Exception) {
                        throw VN97ActivationRecoveryRequired(
                            "prepared activation rollback failed",
                            exc,
                        )
                    }
                    clearPrepared(pending.txId, token)
                }
                VN97BackendTransactionState.ROLLED_BACK ->
                    clearPrepared(pending.txId, token)
            }
        } else {
            if (status.state != VN97BackendTransactionState.ROLLED_BACK) {
                try {
                    backend.rollback(token)
                } catch (exc: Exception) {
                    throw VN97ActivationRecoveryRequired(
                        "rollback recovery failed",
                        exc,
                    )
                }
            }
            finalizeRollback(
                pending.txId,
                VN97InventoryAction.RECOVER_ROLLBACK,
            )
        }
        return inventory()
    }

    private fun clearReserved(txId: String) {
        store.locked { tx ->
            val pending = tx.state.pending
            if (pending != null &&
                pending.txId == txId &&
                pending.phase == "reserved"
            ) {
                tx.state.pending = null
                tx.persist()
            }
        }
    }

    private fun clearPrepared(txId: String, token: String) {
        store.locked { tx ->
            val pending = tx.state.pending
            if (pending != null &&
                pending.txId == txId &&
                pending.backendToken == token
            ) {
                tx.state.pending = null
                tx.persist()
            }
        }
    }

    private fun finalizeActivation(
        txId: String,
        revision: String,
        action: VN97InventoryAction,
    ): VN97CapabilityInventoryItem {
        inventoryRequireRevision(revision, "runtime_revision")
        return store.locked { tx ->
            val state = tx.state
            val pending = state.pending
            if (pending == null ||
                pending.txId != txId ||
                pending.operation != "activate" ||
                pending.phase != "prepared"
            ) {
                throw VN97ActivationRecoveryRequired(
                    "activation transaction changed before finalization"
                )
            }
            val identity = pending.identity
                ?: throw VN97ActivationRecoveryRequired(
                    "prepared activation lacks identity"
                )
            val artifact = pending.artifactSha256
                ?: throw VN97ActivationRecoveryRequired(
                    "prepared activation lacks artifact"
                )
            val token = pending.backendToken
                ?: throw VN97ActivationRecoveryRequired(
                    "prepared activation lacks token"
                )
            val activationId = activationId(
                identity,
                pending.backendId,
                artifact,
                revision,
            )
            val item = VN97CapabilityInventoryItem(
                activationId = activationId,
                capabilityId = identity.capabilityId,
                capabilityVersion = identity.capabilityVersion,
                packageSha256 = identity.packageSha256,
                publisherKeyId = identity.publisherKeyId,
                signatureSha256 = identity.signatureSha256,
                profileId = identity.profileId,
                profileSha256 = identity.profileSha256,
                planSha256 = identity.planSha256,
                runtimeApiVersion = identity.runtimeApiVersion,
                backendId = pending.backendId,
                artifactSha256 = artifact,
                runtimeRevision = revision,
                sourceOrigin = identity.sourceOrigin,
                sourceSha256 = identity.sourceSha256,
                sourceLicense = identity.sourceLicense,
            )
            val stack = state.stacks.getOrPut(item.capabilityId) { mutableListOf() }
            if (stack.size >= 64) {
                throw VN97ActivationRecoveryRequired(
                    "activation stack became full"
                )
            }
            stack += VN97StoredActivation(item, token)
            state.generation += 1L
            state.history += VN97InventoryEvent(
                generation = state.generation,
                action = action,
                capabilityId = item.capabilityId,
                activationId = item.activationId,
                packageSha256 = item.packageSha256,
            )
            if (state.history.size > 4096) {
                repeat(state.history.size - 4096) { state.history.removeAt(0) }
            }
            state.pending = null
            tx.persist()
            item
        }
    }

    private fun finalizeRollback(
        txId: String,
        action: VN97InventoryAction,
    ): VN97CapabilityInventoryItem? = store.locked { tx ->
        val state = tx.state
        val pending = state.pending
        if (pending == null ||
            pending.txId != txId ||
            pending.operation != "rollback"
        ) {
            throw VN97ActivationRecoveryRequired(
                "rollback transaction changed before finalization"
            )
        }
        val stack = state.stacks[pending.capabilityId]
            ?: throw VN97ActivationConflictException(
                "active capability changed during rollback"
            )
        val expectedActivation = pending.activationId
            ?: throw VN97ActivationRecoveryRequired(
                "rollback transaction lacks activation_id"
            )
        if (stack.last().item.activationId != expectedActivation) {
            throw VN97ActivationConflictException(
                "active capability changed during rollback"
            )
        }
        val removed = stack.removeAt(stack.lastIndex)
        if (stack.isEmpty()) state.stacks.remove(removed.item.capabilityId)
        state.generation += 1L
        state.history += VN97InventoryEvent(
            generation = state.generation,
            action = action,
            capabilityId = removed.item.capabilityId,
            activationId = removed.item.activationId,
            packageSha256 = removed.item.packageSha256,
        )
        if (state.history.size > 4096) {
            repeat(state.history.size - 4096) { state.history.removeAt(0) }
        }
        state.pending = null
        tx.persist()
        stack.lastOrNull()?.item
    }

    private fun planIdentity(
        plan: VN97CompatibilityPlan,
        verified: VN97VerifiedCapability,
    ): VN97ActivationIdentity {
        val source = verified.parsed.manifest.source
        return VN97ActivationIdentity(
            capabilityId = plan.capabilityId,
            capabilityVersion = plan.capabilityVersion,
            packageSha256 = plan.packageSha256,
            planSha256 = plan.sha256(),
            profileId = plan.profileId,
            profileSha256 = plan.profileSha256,
            publisherKeyId = plan.publisherKeyId,
            runtimeApiVersion = plan.runtimeApiVersion,
            signatureSha256 = plan.signatureSha256,
            sourceLicense = source.license,
            sourceOrigin = source.origin,
            sourceSha256 = source.sourceSha256,
        )
    }

    private fun activationTransactionId(
        generation: Long,
        identity: VN97ActivationIdentity,
    ): String = hashCanonical(
        VnStrictJson.objectOf(
            "generation" to VnJsonNumber(generation.toString()),
            "identity" to identityToJson(identity),
            "nonce" to VnStrictJson.string(randomNonce()),
            "operation" to VnStrictJson.string("activate"),
        )
    )

    private fun rollbackTransactionId(
        generation: Long,
        activationId: String,
    ): String = hashCanonical(
        VnStrictJson.objectOf(
            "activation_id" to VnStrictJson.string(activationId),
            "generation" to VnJsonNumber(generation.toString()),
            "nonce" to VnStrictJson.string(randomNonce()),
            "operation" to VnStrictJson.string("rollback"),
        )
    )

    private fun activationId(
        identity: VN97ActivationIdentity,
        backendId: String,
        artifactSha256: String,
        revision: String,
    ): String {
        val fields = identityToJson(identity).values.toMutableMap()
        fields["artifact_sha256"] = VnStrictJson.string(artifactSha256)
        fields["backend_id"] = VnStrictJson.string(backendId)
        fields["runtime_revision"] = VnStrictJson.string(revision)
        return hashCanonical(VnJsonObject(fields))
    }

    private fun randomNonce(): String {
        val bytes = ByteArray(16)
        random.nextBytes(bytes)
        return bytes.joinToString("") { "%02x".format(it.toInt() and 0xff) }
    }

    private fun hashCanonical(value: VnJsonValue): String = sha256Hex(
        VnStrictJson.canonical(value).toByteArray(StandardCharsets.UTF_8)
    )
}
